libraries/chain/include/eosio/chain/db_context.hpp

```diff
@@ -0,0 +1,194 @@
+          * interface for uint64_t secondary
+          */
+         virtual int32_t db_idx64_store(uint64_t scope, uint64_t table, account_name payer, uint64_t id,
+                                const uint64_t& secondary) = 0;
+
+         virtual void db_idx64_update(int32_t iterator, account_name payer, const uint64_t& secondary) = 0;
+
+         virtual void db_idx64_remove(int32_t iterator) = 0;
+
+         virtual int32_t
+         db_idx64_find_secondary(uint64_t code, uint64_t scope, uint64_t table, const uint64_t& secondary,
+                                 uint64_t& primary) = 0;
+
+         virtual int32_t db_idx64_find_primary(uint64_t code, uint64_t scope, uint64_t table, uint64_t& secondary,
+                                       uint64_t primary) = 0;
+
+         virtual int32_t db_idx64_lowerbound(uint64_t code, uint64_t scope, uint64_t table, uint64_t& secondary,
+                                     uint64_t& primary) = 0;
+
+         virtual int32_t db_idx64_upperbound(uint64_t code, uint64_t scope, uint64_t table, uint64_t& secondary,
+                                     uint64_t& primary) = 0;
+
+         virtual int32_t db_idx64_end(uint64_t code, uint64_t scope, uint64_t table) = 0;
+
+         virtual int32_t db_idx64_next(int32_t iterator, uint64_t& primary) = 0;
+
+         virtual int32_t db_idx64_previous(int32_t iterator, uint64_t& primary) = 0;
+
+         /**
+          * interface for uint128_t secondary
+          */
+         virtual int32_t db_idx128_store(uint64_t scope, uint64_t table, account_name payer, uint64_t id,
+                                 const uint128_t& secondary) = 0;
+
+         virtual void db_idx128_update(int32_t iterator, account_name payer, const uint128_t& secondary) = 0;
+
+         virtual void db_idx128_remove(int32_t iterator) = 0;
+
+         virtual int32_t
+         db_idx128_find_secondary(uint64_t code, uint64_t scope, uint64_t table, const uint128_t& secondary,
+                                  uint64_t& primary) = 0;
+
+         virtual int32_t db_idx128_find_primary(uint64_t code, uint64_t scope, uint64_t table, uint128_t& secondary,
+                                        uint64_t primary) = 0;
+
+         virtual int32_t db_idx128_lowerbound(uint64_t code, uint64_t scope, uint64_t table, uint128_t& secondary,
+                                      uint64_t& primary) = 0;
+
+         virtual int32_t db_idx128_upperbound(uint64_t code, uint64_t scope, uint64_t table, uint128_t& secondary,
+                                      uint64_t& primary) = 0;
+
+         virtual int32_t db_idx128_end(uint64_t code, uint64_t scope, uint64_t table) = 0;
+
+         virtual int32_t db_idx128_next(int32_t iterator, uint64_t& primary) = 0;
+
+         virtual int32_t db_idx128_previous(int32_t iterator, uint64_t& primary) = 0;
+
+         /**
+          * interface for 256-bit interger secondary
+          */
+         inline static constexpr uint32_t idx256_array_size = 2;
```

Comment:
I think constexpr in this case implies inline so the inline keyword isn't necessary.

Reply:
unneeded copy from another file, removed